p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,388 @@
+		return
+	}
+	rtt := time.Since(tstart)
+
+	if t := msg.GetType(); t != pb.HolePunch_SYNC {
+		s.Reset()
+		hs.handlerError(rp, fmt.Errorf("expected SYNC message from initiator but got %d", t))
+		return
+	}
+	defer s.Close()
+
+	// Hole punch now by forcing a connect
+	pi := peer.AddrInfo{
+		ID:    rp,
+		Addrs: obsDial,
+	}
+	hs.tracer.StartHolePunch(rp, obsDial, rtt)
+	log.Debugf("starting hole punch with %s", rp)
+	counter := atomic.AddInt32(&hs.counter, 1)
+	tstart = time.Now()
+	err := hs.holePunchConnect(pi, int(counter))
+	dt := time.Since(tstart)
+	hs.tracer.EndHolePunch(rp, dt, err)
+	if err != nil {
+		log.Warnf("hole punching with %s failed after %s: %s", rp, dt, err)
+	} else {
+		log.Infof("hole punching with %s successful after %s", rp, dt)
+	}
+}
+
+func (hs *Service) holePunchConnect(pi peer.AddrInfo, attempt int) error {
+	holePunchCtx := network.WithSimultaneousConnect(hs.ctx, "hole-punching")
+	forceDirectConnCtx := network.WithForceDirectDial(holePunchCtx, "hole-punching")
+	dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+	defer cancel()
+
+	hs.tracer.HolePunchAttempt(pi.ID, attempt)
+	err := hs.host.Connect(dialCtx, pi)
+	if err == nil {
+		log.Infof("hole punch with peer %s successful after %d retries; direct conns to peer are:", pi.ID, attempt)
+		for _, c := range hs.host.Network().ConnsToPeer(pi.ID) {
+			if !isRelayAddress(c.RemoteMultiaddr()) {
+				log.Info(c)
+			}
+		}
+	}
+	log.Infof("hole punch attempt %d with peer %s failed: %s; will retry in %s...", attempt, pi.ID, err, retryWait)
+	return err
+}
+
+func isRelayAddress(a ma.Multiaddr) bool {
+	_, err := a.ValueForProtocol(ma.P_CIRCUIT)
+	return err == nil
+}
+
+func addrsToBytes(as []ma.Multiaddr) [][]byte {
+	bzs := make([][]byte, 0, len(as))
+	for _, a := range as {
+		bzs = append(bzs, a.Bytes())
+	}
+	return bzs
+}
+
+func addrsFromBytes(bzs [][]byte) []ma.Multiaddr {
+	addrs := make([]ma.Multiaddr, 0, len(bzs))
+	for _, bz := range bzs {
+		a, err := ma.NewMultiaddrBytes(bz)
+		if err == nil {
+			addrs = append(addrs, a)
+		}
+	}
+	return addrs
+}
+
+type netNotifiee Service
+
+func (nn *netNotifiee) HolePunchService() *Service {
+	return (*Service)(nn)
+}
+
+func (nn *netNotifiee) Connected(_ network.Network, v network.Conn) {
+	hs := nn.HolePunchService()
+	dir := v.Stat().Direction
+
+	// Hole punch if it's an inbound proxy connection.
+	// If we already have a direct connection with the remote peer, this will be a no-op.
+	if dir == network.DirInbound && isRelayAddress(v.RemoteMultiaddr()) {
+		p := v.RemotePeer()
+		hs.activeMx.Lock()
+		_, active := hs.active[p]
+		if !active {
+			hs.active[p] = struct{}{}
+		}
+		hs.activeMx.Unlock()
+
+		if active {
+			return
+		}
+
+		log.Debugf("got inbound proxy conn from peer %s, connectionID is %s", v.RemotePeer().String(), v.ID())
+		hs.refCount.Add(1)
+		go func() {
+			defer hs.refCount.Done()
+			defer func() {
```

Comment:
Combine?

Reply:
We could, but I find it clearer to see what happens first and what happens last this way.